File: frontend/src/services/rag.ts
/**
 * RAG Chatbot service for communicating with the backend API.
 */

import axios from 'axios';

export interface ChatMessage {
  role: 'user' | 'assistant';
  content: string;
}

export interface SourceDocument {
  content: string;
  chapter_id: string;
  section_id: string;
  score: number;
}

export interface ChatRequest {
  message: string;
  conversation_history: ChatMessage[];
  chapter_id?: string;
  section_id?: string;
  selected_text?: string;
}

export interface ChatResponse {
  answer: string;
  sources: SourceDocument[];
  conversation_history: ChatMessage[];
}

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000/api';

const api = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

/**
 * Send a chat message to the RAG chatbot.
 */
export async function sendChatMessage(request: ChatRequest): Promise<ChatResponse> {
  const response = await api.post<ChatResponse>('/rag/chat', request);
  return response.data;
}

/**
 * Index a document chunk for RAG search.
 */
export async function indexDocument(content: string, chapterId: string, sectionId?: string): Promise<void> {
  await api.post('/rag/index', {
    content,
    chapter_id: chapterId,
    section_id: sectionId,
  });
}

/**
 * Index multiple document chunks in batch.
 */
export async function indexDocumentsBatch(documents: Array<{
  content: string;
  chapter_id: string;
  section_id?: string;
}>): Promise<void> {
  await api.post('/rag/index/batch', { documents });
}

/**
 * Check the health of the RAG service.
 */
export async function checkRagHealth(): Promise<{
  status: string;
  qdrant_connected: boolean;
  openai_connected: boolean;
}> {
  const response = await api.get('/rag/health');
  return response.data;
}

File: backend/src/schemas/rag.py
"""
RAG chatbot schemas.

This module defines request/response models for the RAG chatbot API.
"""

from typing import List, Optional

from pydantic import BaseModel, Field


class ChatMessage(BaseModel):
    """A single message in the chat conversation."""

    role: str = Field(..., description="Message role: 'user' or 'assistant'")
    content: str = Field(..., description="Message content")


class ChatRequest(BaseModel):
    """Request model for chat endpoint."""

    message: str = Field(..., description="User's question or message")
    conversation_history: List[ChatMessage] = Field(
        default_factory=list,
        description="Previous conversation messages",
    )
    chapter_id: Optional[str] = Field(
        None,
        description="Optional chapter ID to scope the search",
    )
    section_id: Optional[str] = Field(
        None,
        description="Optional section ID to scope the search",
    )
    selected_text: Optional[str] = Field(
        None,
        description="Optional text selected by user for context",
    )


class SourceDocument(BaseModel):
    """A source document chunk used in the response."""

    content: str = Field(..., description="Document content")
    chapter_id: str = Field(..., description="Chapter identifier")
    section_id: str = Field(..., description="Section identifier")
    score: float = Field(..., description="Relevance score")


class ChatResponse(BaseModel):
    """Response model for chat endpoint."""

    answer: str = Field(..., description="AI-generated answer")
    sources: List[SourceDocument] = Field(
        default_factory=list,
        description="Source documents used in the answer",
    )
    conversation_history: List[ChatMessage] = Field(
        default_factory=list,
        description="Updated conversation history",
    )


class IndexRequest(BaseModel):
    """Request model for indexing documents."""

    content: str = Field(..., description="Document content to index")
    chapter_id: str = Field(..., description="Chapter identifier")
    section_id: Optional[str] = Field(None, description="Section identifier")
    metadata: Optional[dict] = Field(
        None, description="Additional metadata"
    )


class IndexBatchRequest(BaseModel):
    """Request model for batch indexing documents."""

    documents: List[IndexRequest] = Field(
        ..., description="List of documents to index"
    )


class IndexResponse(BaseModel):
    """Response model for indexing endpoints."""

    success: bool = Field(..., description="Whether indexing succeeded")
    chunk_ids: List[str] = Field(
        default_factory=list, description="IDs of indexed chunks"
    )
    message: str = Field(..., description="Status message")


class HealthResponse(BaseModel):
    """Response model for health check endpoint."""

    status: str = Field(..., description="Service status")
    qdrant_connected: bool = Field(
        ..., description="Whether Qdrant is connected"
    )
    openai_connected: bool = Field(
        ..., description="Whether OpenAI is accessible"
    )
